### ISSUE TITLE
Unittest with system args in Cement 3.0

### ISSUE THREAD

**Issue by unleashalicia** [datafolklabs/cement#557]
**System Information**

- Cement Version: 3.0
- Python Version: 3.6.0
- Operating System and Version: Mac OS 10.13.6

I am using Nose tests on my application.  I have a test I would like to write that is in my main application file and takes command line arguments.  There is no documentation on cement 3.0 testing: https://docs.builtoncement.com/additional-topics/unit-testing

I tried to use cement 2's docs, but they are not the same: https://cement.readthedocs.io/en/2.10/dev/testing/

The most immediate error I get (in my linter) is "Module 'cement.utils.test' has no 'CementTestCase'".

Do you have any feedback or sources on how I can set up a unittest with command line arguments using cement 3.0?

**Comment by derks** [comment]
@unleashalicia thank you for the reminder that I never completed that section of the Cement 3.0 docs! I apologize for that ... it's been long enough, a "FIXME" doesn't really save me anymore. :0

Cement 3 has changed with regards to testing... but is similar.  There is now a `cement.core.foundation.TestApp` that is setup for testing:

- https://cement.readthedocs.io/en/3.0/api/core/foundation/#cement.core.foundation.TestApp

Admittedly, the API doc is no more helpful either.  :/

That said, if you use the `cement generate project` utility you should already have a `MyAppTest`

**myapp/main.py**

```python
from cement import App, TestApp

class MyApp(App):
    ...

class MyAppTest(TestApp, MyApp):
   ...
```

When subclassing from `TestApp` you are zeroing out some key meta options that are likely to interfere with testing like `config_files`, etc.  The generated project also includes testing with argument examples:

**tests/test_myapp.py**:

```pytohn
from myapp.main import MyAppTest

def test_myapp():
    # test myapp without any subcommands or arguments
    with MyAppTest() as app:
        app.run()
        assert app.exit_code == 0


def test_myapp_debug():
    # test that debug mode is functional
    argv = ['--debug']
    with MyAppTest(argv=argv) as app:
        app.run()
        assert app.debug is True
```

You pass arguments via `App.meta.argv` or the `argv` keyword argument.  Does that help?

**Comment by derks** [comment]
This has been documented, and will be released with Cement 3.0.8:

- https://docs.builtoncement.com/additional-topics/unit-testing
